baseplate/events/publisher.py

```diff
@@ -0,0 +1,203 @@
+        if not self.batch_start:
+            self.batch_start = time.time()
+
+    def flush(self):
+        """Explicitly flush the batch if any items are enqueued."""
+        if self.batch:
+            self.consumer.consume_batch(self.batch)
+            self.batch = []
+            self.batch_size = self.consumer.batch_size_overhead
+            self.batch_start = None
+
+
+def gzip_compress(content):
+    buf = BytesIO()
+    with gzip.GzipFile(fileobj=buf, mode="wb", compresslevel=9) as f:
+        f.write(content)
+    return buf.getvalue()
+
+
+class BatchConsumer(object):
+    batch_size_overhead = 2  # the [] that wrap the json list
+    batch_size_limit = 500 * 1024
+
+    def __init__(self, metrics_client, cfg):
+        self.metrics = metrics_client
+        self.url = "https://%s/v1" % cfg.collector.hostname
+        self.key_name = cfg.key.name
+        self.key_secret = cfg.key.secret
+        self.session = requests.Session()
+
+    @staticmethod
+    def get_item_size(item):
+        # the item is already serialized, so we'll just add one byte for comma
+        return len(item) + 1
+
+    def _sign_payload(self, payload):
+        digest = hmac.new(self.key_secret, payload, hashlib.sha256).hexdigest()
+        return "key={key}, mac={mac}".format(key=self.key_name, mac=digest)
+
+    def consume_batch(self, items):
+        logger.info("sending batch of %d items", len(items))
+        payload = b"[" + b",".join(items) + b"]"
+        compressed_payload = gzip_compress(payload)
+        headers = {
+            "Date": email.utils.formatdate(usegmt=True),
+            "User-Agent": "baseplate-event-publisher/1.0",
+            "Content-Type": "application/json",
+            "X-Signature": self._sign_payload(payload),
+            "Content-Encoding": "gzip",
+        }
+
+        while True:
+            try:
+                with self.metrics.timer("post"):
+                    response = self.session.post(
+                        self.url,
+                        headers=headers,
+                        data=compressed_payload,
+                        timeout=POST_TIMEOUT,
+
+                        # http://docs.python-requests.org/en/latest/user/advanced/#keep-alive
+                        stream=False,
+                    )
+                response.raise_for_status()
+            except requests.HTTPError as exc:
+                self.metrics.counter("error.http").increment()
+                logger.exception("HTTP Request failed")
+
+                # we should crash if it's our fault
+                if getattr(exc, "response", None) and exc.response.status_code < 500:
+                    raise
+            except IOError:
+                self.metrics.counter("error.io").increment()
+                logger.exception("HTTP Request failed")
+            else:
+                self.metrics.counter("sent").increment(len(items))
+                return
+
+            time.sleep(RETRY_DELAY)
```

Comment:
:no_bell: OK I'm just gonna think through this out loud, I think I've got it but I want to make sure there are no queue sharing issues and maybe it'll be helpful to someone else later:

Each consumer daemon has a single batcher. Each batcher has a maximum size and age.

The daemon will fill the batcher with messages it consumes from the queue.

If the batcher reaches the maximum size or age, it will flush synchronously. That flush is an http POST. If that flush returns a 500, it will sleep, halting the consumer, and retry in 2 seconds. As this happens all other consumers are free to batch as they like.

They may consume and send just fine, which means that events may be sent out of order. (Which is fine and we should expect).

If the event collector starts returning 500s for everyone, every consumer daemon will reach this state, retrying every 2 seconds. This means that nothing will be consuming from the top level queues, but they still may be added to. The event collector is likely receiving POSTs on success from every consumer every second or two anyhow, so this load will be expected. A fixed retry of 2seconds is probably fine.

If we've been in a 500 state for a minute or two and have a whole bunch of full queues, we'll start consuming them and batching them right away as we hit max size. This may overwhelm the event collector server, but if it does it will start to return 500 level responses (or timeouts) and we'll retry in two seconds, effectively slowing us down.

OK, phew. I think I now understand this better and this feels great. :+1: